[concurrency] allow type initializers to be 'async'

kavon opened the pull request:
Earlier in the design of async/await, we disallowed async initializers for nominal types. There is no fundamental reason for the restriction and it was decided that async initializers would actually be useful (to allow initializers to invoke async functions, for example), and so they appear in the now accepted async/await proposal.

This PR fills in this missing part of the proposal, by allowing initializers to be async.

Resolves rdar://71049563

TODO:
- [x] Lift restriction in the parser
- [x] Implement checking missing explicit call to async, zero-arg, designated super initializer
- [x] Update all existing tests
- [x] Ensure that asynchronous global-actor-isolated initializers emit the appropriate actor hop.
- [x] Test coverage
  - [x] typechecker
  - [x] SILGen
  - [x] execution
  - [x] serialization

kavon committed:
allow type initializers to be 'async'

implicit calls to an async super.init are not allowed

kavon committed:
update existing regression tests after small diagnostics refactor

new phrasing in the error messages to match up with
other phrasing used in messages.

kavon committed:
add new tests for async initializers

kavon commented:
@swift-ci please test

DougGregor commented on lib/ClangImporter/ImportName.cpp:
```diff
@@ -1286,6 +1286,7 @@ NameImporter::considerAsyncImport(
 
   // Initializers cannot be 'async'.
   // FIXME: We might eventually allow this.
+  // TODO: should the restriction be lifted in ClangImporter?
```
Hmm. I think you might need to implement the corresponding restriction on the Swift side to prevent `@objc` `async` initializers, e.g., make sure that we reject

```swift
class X: NSObject {
  @objc init() async { ... }
}
```

because we can no longer rely on the the `async` init to be rejected.

DougGregor reviewed (approved):
Looking very good! I'm glad that the code changes were relatively modest.

DougGregor commented on test/Concurrency/async_initializer.swift:
```diff
@@ -0,0 +1,138 @@
+////
+// test other types with constructors
+
+actor A {
```
Could you also add some examples with global actors and async initializers?

kavon replied:
👍🏼 done. thanks for pointing this out

kavon committed:
[nfc] rename one of the two `emitProlog`s in SILGen

There's a basic prolog emission function, used by value and class constructors, etc, and then there's the full-blown one for functions and closures, which uses the basic version.

kavon committed:
ensure that async initializers for reference and value
types correctly hop to the right global-actor in their
prolog.

kavon commented:
@swift-ci Please smoke test and merge

kavon commented:
@swift-ci please smoke test and merge
